File: Packages/qb-taxijob/Client/html/meter.js
let meterStarted = false;

const updateMeter = (currentFare, distanceTraveled) => {
    $("#total-price").html("$ " + currentFare.toFixed(2));
    $("#total-distance").html(distanceTraveled.toFixed(2) + " mi");
};

const resetMeter = () => {
    $("#total-price").html("$ 0.00");
    $("#total-distance").html("0.00 mi");
};

const toggleMeter = (enabled) => {
    if (enabled) {
        $(".toggle-meter-btn").html("<p>Started</p>");
        $(".toggle-meter-btn p").css({ color: "rgb(51, 160, 37)" });
    } else {
        $(".toggle-meter-btn").html("<p>Stopped</p>");
        $(".toggle-meter-btn p").css({ color: "rgb(231, 30, 37)" });
    }
};

const meterToggle = () => {
    if (!meterStarted) {
        $.post(
            "https://qb-taxijob/enableMeter",
            JSON.stringify({
                enabled: true,
            })
        );
        toggleMeter(true);
        meterStarted = true;
    } else {
        $.post(
            "https://qb-taxijob/enableMeter",
            JSON.stringify({
                enabled: false,
            })
        );
        toggleMeter(false);
        meterStarted = false;
    }
};

const openMeter = (startingPrice) => {
    $(".container").fadeIn(150);
    $("#total-price-per-100m").html("$ " + startingPrice.toFixed(2));
};

const closeMeter = () => {
    $(".container").fadeOut(150);
};

$(".container").hide();

Events.Subscribe('TOGGLE_METER_UI', (enabled, startingPrice) => {
    if (enabled) {
        openMeter(startingPrice);
    } else {
        closeMeter();
    }
})

Events.Subscribe('TOGGLE_METER_COUNTING', () => {
    meterToggle()
})

Events.Subscribe('UPDATE_METER', (currentFare, distanceTraveled) => {
    updateMeter(currentFare, distanceTraveled)
})

Events.Subscribe('RESET_METER', () => {
    resetMeter();
})

File: Packages/qb-taxijob/Client/meter.lua
--local meterUI = WebUI('meter', 'file://html/meter.html')
local previousCoord = nil
local meterData = {
    enabled = false,
    currentFare = 0,
    distanceTraveled = 0,
    timer = nil,
}

-- Functions

local function EnableMeter()
    meterData.timer = Timer.SetInterval(function()
        local ped = Client.GetLocalPlayer():GetControlledCharacter()
        if not ped then return end
        
        local pedLocation = ped:GetLocation()
        previousCoord = previousCoord or pedLocation
        meterData.distanceTraveled = meterData.distanceTraveled + previousCoord:Distance(pedLocation)
        meterData.currentFare = math.floor(meterData.distanceTraveled * Config.Meter.Rate + Config.Meter.StartingPrice)
        previousCoord = pedLocation
        meterUI:CallEvent('UPDATE_METER', meterData.currentFare, meterData.distanceTraveled)
    end, 4000)
end

-- Handlers

Events.Subscribe('qb-taxijob:client:toggleMeter', function()
    meterUI:CallEvent('TOGGLE_METER_UI', Config.Meter.StartingPrice)
end)

Events.Subscribe('qb-taxijob:client:enableMeter', function()
    meterData.enabled = not meterData.enabled
    if not meterData.enabled then 
        Timer.ClearInterval(meterData.timer)
        meterData = {
            enabled = false,
            currentFare = 0,
            distanceTraveled = 0,
            timer = nil,
        }
        meterUI:CallEvent('RESET_METER')
        return 
    end
    EnableMeter()
end)

